src/cargo_communication/request_actor.rs

```diff
@@ -0,0 +1,350 @@
+                        parents: vec![test_task_id.id.clone()],
+                    },
+                    suite_task_progress: SuiteTaskProgress {
+                        progress: 0,
+                        total: 0,
+                    },
+                    test_task_id,
+                    single_test_task_ids: HashMap::new(),
+                })
+            }
+            _ => ExecutionState::Compile,
+        }
+    }
+}
+
+impl<R, C> RequestActor<R, C>
+where
+    R: Request,
+    R::Params: CreateCommand + CreateResult<R::Result>,
+    C: CargoHandleTrait<CargoMessage>,
+{
+    pub fn new(
+        sender: Box<dyn Fn(RPCMessage) + Send>,
+        req_id: RequestId,
+        params: R::Params,
+        root_path: &Path,
+    ) -> RequestActor<R, C> {
+        RequestActor {
+            sender,
+            cargo_handle: None,
+            req_id,
+            state: RequestActorState::new::<R>(params.origin_id()),
+            params,
+            root_path: root_path.to_path_buf(),
+        }
+    }
+
+    fn next_event(&self, inbox: &Receiver<Event>) -> Option<Event> {
+        let cargo_chan = self.cargo_handle.as_ref().map(|cargo| cargo.receiver());
+        select! {
+            recv(inbox) -> msg => msg.ok(),
+            recv(cargo_chan.unwrap_or(&never())) -> msg => match msg {
+                Ok(msg) => Some(Event::CargoEvent(msg)),
+                Err(_) => Some(Event::CargoFinish),
+            }
+        }
+    }
+
+    pub fn spawn_handle(&mut self) -> Result<CargoHandle, String> {
+        let command = self.params.create_command(self.root_path.clone());
+        info!("Created command: {:?}", command);
+        match CargoHandle::spawn(command) {
+            Ok(cargo_handle) => Ok(cargo_handle),
+            Err(err) => {
+                todo!()
+            }
+        }
+    }
+
+    pub fn run(mut self, cancel_receiver: Receiver<Event>, cargo_handle: C) {
+        self.cargo_handle = Some(cargo_handle);
+
+        self.report_task_start(self.state.root_task_id.clone(), None, None);
+        self.start_compile_task();
+
+        while let Some(event) = self.next_event(&cancel_receiver) {
+            match event {
+                Event::Cancel => {
+                    self.cancel();
+                    return;
+                }
+                Event::CargoFinish => {
+                    self.finish_command();
+                    return;
+                }
+                Event::CargoEvent(message) => {
+                    // handle information and create notification based on that
+                    match message {
+                        CargoMessage::CargoStdout(stdout) => self.handle_cargo_information(stdout),
+                        CargoMessage::CargoStderr(stderr) => {
+                            self.log_message(MessageType::Error, stderr, None)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn start_compile_task(&mut self) {
```

Comment:
Przeniósłbym tą funkcję do cargo_messages_handle

Reply:
Założeniem cargo_messages_handle było to, żeby znajdowały się tam funkcje dotyczące przetwarzania wiadomości od cargo (i tworzenia odpowiednich notification na ich podstawie). Ta funkcja nie dotyczy wiadomości od cargo, więc myślę że nie ma sensu jej tam przenosić